### PR TITLE
Mbedtls tfm compat

### PR DESCRIPTION
**Pull request by joerchan**
## Description

Fix mbedtls compatibility issues with the TF-M (trusted-firmware-m) PSA API headers.
The TF-M and Mbed TLS have their own set of PSA API headers, which leads to some problems.

When mbedtls (TLS library) is compiled using the Mbed TLS PSA API headers using TrustZone and TF-M we get a link error, because of the missing function psa_set_key_domain_parameters referred to from psa_set_key_type (psa/crypto_struct.h).
When using the trusted-firmware-m PSA API headers mbedtls ends up with a lot of errors. The psa_crypto_core.h header file has a lot of undedefined structs types, and the mbedtls_ecc helper functions are missing.
These two commits fixes these issues.

## PR checklist

- [x] **changelog** provided
- [x] **backport** no
- [x] **tests** just the CI

### CODE REVIEW

**Comment by joerchan** [comment]
@valeriosetti Fixes issues in TF-M 2.0.0 and Mbed TLS 3.5.0

**Comment by frkv** [comment]
This is aligned with ongoing discussion on helper-functions: https://github.com/Mbed-TLS/mbedtls/issues/7764 

Putting these in psa_util.h/c is aligned with this comment: https://github.com/Mbed-TLS/mbedtls/issues/7764#issuecomment-1849707955

**Comment by mpg** [comment]
I have nothing against the changes in his PR (as Frank notes, I've commented about one of them earlier), but I'd like to better understand why they are needed, in order to avoid similar issues in the future.

> When mbedtls (TLS library) is compiled using the Mbed TLS PSA API headers using TrustZone and TF-M we get a link error, because of the missing function psa_set_key_domain_parameters referred to from psa_set_key_type (psa/crypto_struct.h).
> When using the trusted-firmware-m PSA API headers mbedtls ends up with a lot of errors. The psa_crypto_core.h header file has a lot of undedefined structs types, and the mbedtls_ecc helper functions are missing.

Could you give more details so that we can reproduce? What do you compile "using the Mbed TLS PSA API headers using TrustZone and TF-M" and how (configuration steps etc)?

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
This pull request moves two functions, both the declarations and the implementations. This move makes sense: the functions are bridges between the PSA interface and the mbedtls interface, so `psa/crypto_extra.h` wasn't a good fit, `mbedtls/psa_util.h` is a better fit.

In terms of backward compatibility, `#include <psa/crypto.h>` no longer declares the moved functions: application code now has to use `#include <mbedtls/psa_util.h>`. This would normally be a backward compatibility break, but the functions are explicitly noted as “may be removed at any time without notice”, so that's acceptable.

So I'm approving the change, but please add a [changelog entry](https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md) under “Changes” to indicate that the functions have moved.

**Comment by gilles-peskine-arm** [comment]
And please fix the build (see CI results). There are source files in the project that now need an extra `#include`.

**Comment by gilles-peskine-arm** [comment]
Priority justification: high rather than low/medium because this comes from a TF-M product requirement.

**Comment by joerchan** [comment]
@mpg It's not the easiest thing to reproduce, as we need need to build mbedtls ourselves to make it work with TF-M.

If you want to check this out and are willing to setup a different build environment then zephyr TF-M PSA Crypto sample can be used.
https://docs.zephyrproject.org/latest/samples/tfm_integration/psa_crypto/README.html

This sample broke when we upgraded to TF-M 1.8.0 and Mbed TLS 3.5.0.
This sample worked when built using TF-M 1.7.0 and Mbed TLS 3.4.0 however it was not really built properly, as it was using the Mbed TLS PSA headers, and not the TF-M PSA headers.

To be clear on what I mean, this are the different PSA headers:

Mbed TLS PSA headers
https://github.com/Mbed-TLS/mbedtls/tree/v3.5.0/include/psa

TF-M PSA headers.
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/interface/include/psa?h=release/2.0.x

These API headers are not interchangeable, so care must be taken to build with the right set.
Example of what can go wrong:

Mbed TLS PSA:
```
static inline void psa_set_key_type( psa_key_attributes_t *attributes,
                                    psa_key_type_t type )
{
    if( attributes->MBEDTLS_PRIVATE(domain_parameters) == NULL )
    {
        /* Common case: quick path */
        attributes->MBEDTLS_PRIVATE(core).MBEDTLS_PRIVATE(type) = type;
    }
    else
    {
        /* Call the bigger function to free the old domain paramteres.
         * Ignore any errors which may arise due to type requiring
         * non-default domain parameters, since this function can't
         * report errors. */
        (void) psa_set_key_domain_parameters( attributes, type, NULL, 0 );
    }
}
```
TF-M PSA:

```
static inline void psa_set_key_type(psa_key_attributes_t *attributes,
                                    psa_key_type_t type)
{
    attributes->type = type;
}
```

If using TF-M and the Mbed TLS PSA header, you will now get an undefined reference to psa_set_key_domain_parameters.

To build and use TF-M we need to make sure that we have the right include path order, and that we don't include the psa implementation in the mbedcrypto library.

Set the include ordel to prioritize TF-M PSA over MBed TLS PSA.
```
    if (CONFIG_BUILD_WITH_TFM)
      target_include_directories(mbedTLS INTERFACE
        $<TARGET_PROPERTY:tfm,TFM_BINARY_DIR>/api_ns/interface/include
      )
    endif()

    # Add regular includes
    target_include_directories(mbedTLS INTERFACE
        ${ZEPHYR_CURRENT_MODULE_DIR}/include
        configs
        include
    )
```

Exclude the PSA implementation sources when build with TF-M.
```
    if (CONFIG_MBEDTLS_PSA_CRYPTO_C AND NOT CONFIG_BUILD_WITH_TFM)
      list(APPEND crypto_source
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_aead.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_cipher.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_driver_wrappers_no_static.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_ecp.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_hash.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_mac.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_rsa.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_se.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_storage.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_its_file.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_client.c
        ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_slot_management.c
      )
    endif()
```

With the include of psa_crypto_core.h header included from psa_util.h I get these errors:
```
mbedtls/library/psa_util.c:27:
mbedtls/library/psa_crypto_core.h:45:5: error: unknown type name 'psa_core_key_attributes_t'
   45 |     psa_core_key_attributes_t attr;
```

And with the mbedtls_ecc helper functions in psa_crypto.c and psa/crypto_extra.h I get implicit declaration and undefined reference to the mbedtls_ecc functions.
Because TF-M psa/crypto_extra.h does not have have the definition of these functions, and I have excluded psa_crypto.c from the build.

```
library/pk_internal.h:99:14: warning: implicit declaration of function 'mbedtls_ecc_group_of_psa';
  did you mean 'mbedtls_ecp_group_copy'? [-Wimplicit-function-declaration]
   99 |         id = mbedtls_ecc_group_of_psa(curve, psa_get_key_bits(&opaque_attrs), 0);
      |              ^~~~~~~~~~~~~~~~~~~~~~~~
      |              mbedtls_ecp_group_copy
...
mbedtls/library/pk_internal.h:99: undefined reference to `mbedtls_ecc_group_of_psa'
```

File: ChangeLog.d/move-mbedtls-ecc-psa-helpers [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Changes
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/move-mbedtls-ecc-psa-helpers:1]: The file needs to have a `.txt` extension

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by mpg** [comment]
@joerchan Thank you very much for the detailed explanation of the problem and steps to reproduce. As you noted, this is not entirely straightforward, so might not have time to go through the process right now, but I'm keeping a link to your comment and I'll get there eventually.

**Pull request merged**